example-apps/chatbot-rag-app/data/index_data.py

```diff
@@ -82,20 +110,86 @@ def main():
 
     print(f"Split {len(workplace_docs)} documents into {len(docs)} chunks")
 
-    print(f"Creating Elasticsearch sparse vector store in {ELASTICSEARCH_URL}")
+    print(f"Creating Elasticsearch sparse vector store for {ELASTICSEARCH_URL}")
 
-    elasticsearch_client.indices.delete(index=INDEX, ignore_unavailable=True)
-
-    ElasticsearchStore.from_documents(
-        docs,
-        es_connection=elasticsearch_client,
+    store = ElasticsearchStore(
+        es_connection=es,
         index_name=INDEX,
         strategy=ElasticsearchStore.SparseVectorRetrievalStrategy(model_id=ELSER_MODEL),
-        bulk_kwargs={
-            "request_timeout": 60,
-        },
     )
 
+    # The first call creates ML tasks to support the index, and typically fails
+    # with the default 10-second timeout, at least when Elasticsearch is a
+    # container running on Apple Silicon.
+    #
+    # Once elastic/elasticsearch#107077 is fixed, we can use bulk_kwargs to
+    # adjust the timeout.
+
+    print(f"Adding documents to index {INDEX}")
+
+    spinner = Halo(text="Processing bulk operation", spinner="dots")
+    if stdout.isatty():
+        spinner.start()
+
+    try:
+        es.indices.delete(index=INDEX, ignore_unavailable=True)
+        store.add_documents(list(docs))
+    except BadRequestError:
+        # This error means the index already exists
+        pass
+    except (ConnectionTimeout, ApiError) as e:
+        if isinstance(e, ApiError) and e.status_code != 408:
+            raise
+        warn(f"Error occurred, will retry after ML jobs complete: {e}")
+        await_ml_tasks()
+        es.indices.delete(index=INDEX, ignore_unavailable=True)
+        store.add_documents(list(docs))
+
+    if stdout.isatty():
+        spinner.stop()
+
+    print(f"Documents added to index {INDEX}")
+
+
+def await_ml_tasks(max_timeout=600, interval=5):
+    """
+    Waits for all machine learning tasks to complete within a specified timeout period.
+
+    Parameters:
+        max_timeout (int): Maximum time to wait for tasks to complete, in seconds.
+        interval (int): Time to wait between status checks, in seconds.
+
+    Raises:
+        TimeoutError: If the timeout is reached and machine learning tasks are still running.
+    """
+    start_time = time.time()
+    ml_tasks = get_ml_tasks()
+    if not ml_tasks:
+        return  # likely a lost race on tasks
+
+    print(f"Awaiting {len(ml_tasks)} ML tasks")
+
+    while time.time() - start_time < max_timeout:
+        ml_tasks = get_ml_tasks()
+        if len(ml_tasks) == 0:
+            break
+        time.sleep(interval)
+
+    if ml_tasks:
```

Comment:
Can remove this `if` with above `return`, just small simplification

Reply:
well if we removed this part, we would pass even if ML tasks are still processing, which creates a worse timeout later, right?

Reply:
sorry forgot to refresh. good idea